### DISCUSSION TITLE
Rationale for removing namespace customization and amalgamated header?

### DISCUSSION THREAD

**Discussion by gitamohr** [taocpp/PEGTL#346] [Q&A] [answered]
These two features were important to our decision to adopt PEGTL, since they made it easy to embed in our project.  I'm curious what was the motivation for removing these features?

**Comment by ColinH** [comment, marked as answer]
The amalgamated header was removed because (a) we don't see much value in it, if you can copy one header you can copy a directory of them (for us the dividing line for ease-of-use is header-only vs. needs-to-be-compiled), (b) there are several choices of what all to include (no pun intended) or not; if `contrib` is part of the amalgamated header then you pull in ICU (!) and `iostream` as dependencies which not everybody will want ... and (c) it was more work to maintain than one might think.

The namespace customisation was originally introduced shortly after placing a copy of the PEGTL in [taoJSON](https://github.com/taocpp/json), i.e. to prevent a clash when a project uses both the PEGTL and taoJSON where the latter embeds a different version of the former. Now that we are using submodules it didn't seem to be necessary anymore. What is your use case for this feature?

**Reply by gitamohr** [reply]
Thank you for the reply.  I believe our use case is similar.  I work on Pixar's [Universal Scene Description (OpenUSD)](https://openusd.org/release/index.html).  We're embedding PEGTL into OpenUSD.  OpenUSD in turn is embedded into several other software systems; for example, Autodesk's Maya, SideFX's Houdini, NVIDIA's Omniverse, Apple's ARKit, Unreal, Unity, and so on.  OpenUSD itself supports versioned & customized namespaces so that multiple copies can coexist within a single process.  So in order to embed the PEGTL safely, we must be able to similarly namespace it to avoid possible conflicts with any other in-process PEGTLs.

**Reply by ColinH** [reply]
To be clear, you are tracking the main branch rather than using the latest 3.x release?

**Reply by ColinH** [reply]
You have convinced us that the namespace customisation was a useful feature beyond our immediate use case and will re-introduce it.

**Reply by gitamohr** [reply]
No we are currently working with the released versions. This question is forward-looking; I happened to notice that these were removed from the main branch. There's no urgency here on our end; it would just be a barrier for us to upgrade to a future 4.x version.

**Reply by gitamohr** [reply]
That's great, thank you very much indeed!

**Reply by ColinH** [reply]
Can you continue to work with the PEGTL if we don't make an amalgamated header again?

**Reply by gitamohr** [reply]
With the ability to customize the namespace, I _think_ so. I can try to see what our embedding would look like without an amalgamated header and report back if I encounter any problems.

**Reply by ColinH** [reply]
The customisable namespace is back in the main branch, hopefully 100% compatible to the previous version. We'll continue to hope that the missing amalgamated header won't be a problem.

**Reply by gitamohr** [reply]
We just moved from 2.x to 3.x, and in that process we switched away from using the amalgamated header with no problems.  The namespace customization is working perfectly.  Thanks again!
